lib/ruby_smb/gss/provider/ntlm.rb

```diff
@@ -0,0 +1,337 @@
+      # A GSS provider that authenticates clients via the NT LAN Manager (NTLM) Security Support Provider (NTLMSSP)
+      # protocol.
+      #
+      class NTLM < Base
+        # An account representing an identity for which this provider will accept authentication attempts.
+        Account = Struct.new(:username, :password, :domain) do
+          def to_s
+            "#{domain}\\#{username}"
+          end
+        end
+
+        class OSVersion < BinData::Record
+          endian :big
+
+          uint8  :major
+          uint8  :minor
+          uint16 :build
+          uint32 :ntlm_revision, initial_value: 15
+
+          def to_s
+            "Version #{major}.#{minor} (Build #{build}); NTLM Current Revision #{ntlm_revision}"
+          end
+        end
+
+        class Authenticator < Authenticator::Base
+          NEGOTIATE_FLAGS = {
+            :UNICODE                  => 1 << 0,
+            :OEM                      => 1 << 1,
+            :REQUEST_TARGET           => 1 << 2,
+            :SIGN                     => 1 << 4,
+            :SEAL                     => 1 << 5,
+            :DATAGRAM                 => 1 << 6,
+            :LAN_MANAGER_KEY          => 1 << 7,
+            :NTLM                     => 1 << 9,
+            :NT_ONLY                  => 1 << 10,
+            :ANONYMOUS                => 1 << 11,
+            :OEM_DOMAIN_SUPPLIED      => 1 << 12,
+            :OEM_WORKSTATION_SUPPLIED => 1 << 13,
+            :ALWAYS_SIGN              => 1 << 15,
+            :TARGET_TYPE_DOMAIN       => 1 << 16,
+            :TARGET_TYPE_SERVER       => 1 << 17,
+            :TARGET_TYPE_SHARE        => 1 << 18,
+            :EXTENDED_SECURITY        => 1 << 19,
+            :IDENTIFY                 => 1 << 20,
+            :NON_NT_SESSION           => 1 << 22,
+            :TARGET_INFO              => 1 << 23,
+            :VERSION_INFO             => 1 << 25,
+            :KEY128                   => 1 << 29,
+            :KEY_EXCHANGE             => 1 << 30,
+            :KEY56                    => 1 << 31
+          }.freeze
+
+          def reset!
+            super
+            @server_challenge = nil
+          end
+
+          def process(request_buffer=nil)
+            if request_buffer.nil?
+              # this is only NTLMSSP (as opposed to SPNEGO + NTLMSSP)
+              buffer = OpenSSL::ASN1::ASN1Data.new([
+                Gss::OID_SPNEGO,
+                OpenSSL::ASN1::ASN1Data.new([
+                  OpenSSL::ASN1::Sequence.new([
+                    OpenSSL::ASN1::ASN1Data.new([
+                      OpenSSL::ASN1::Sequence.new([
+                        Gss::OID_NTLMSSP
+                      ])
+                    ], 0, :CONTEXT_SPECIFIC),
+                    OpenSSL::ASN1::ASN1Data.new([
+                      OpenSSL::ASN1::ASN1Data.new([
+                        OpenSSL::ASN1::ASN1Data.new([
+                          OpenSSL::ASN1::GeneralString.new('not_defined_in_RFC4178@please_ignore')
```

Comment:
I get it's a niche edgecase, but would it be worth changing this to something that an IDS will be less likely to pick up on. Could be mistaken but I'm sure I saw `not_defined_in_RFC4178@please_ignore` in a packet and it stuck out like a sore thumb.

Reply:
That's actually from the MS docs, see [[MS-SPNG] section 2.2.1 NegTokenInit2:](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-spng/8e71cf53-e867-4b79-b5b5-38c92be3d472)

> hintName: SHOULD<5> contain the string "not_defined_in_RFC4178@please_ignore".

Reply:
Kewl, if it's the default value then it no worries